Implement RawAHashTable and AHashMap and AHashSet based on it

Brogolem35 opened the pull request:
This PR adds RawAHashTable abstract class and bases the existing AHashMap on top of it. An AHashSet implementation is also added that is based on the RawAHashTable. AHashSet is a drop in replacement to the existing HashSet implementation.

This is done to lessen the issues mentioned in the godotengine/godot-proposals#12515 and is an execution of my [proposed solution](https://github.com/godotengine/godot-proposals/issues/12515#issuecomment-3084266986) on the same thread.

This PR does not change any of the existing uses of the HashSet, except for the one in the `editor/animation/animation_blend_tree_editor_plugin.cpp`. A branch that changes all occurances of HashSet to AHashSet can be found [here](https://github.com/Brogolem35/godot/tree/rawahashtable_all) (may not compile in non-Linux platforms).

# Differences with HashSet

|                  | HashSet             | AHashSet            |
| ---------------- | ------------------- | ------------------- |
| Size             | Prime numbers       | Powers of 2         |
| Wrapping         | Fastmod or if check | Masking             |
| Metadata Storage | Structure of Arrays | Array of Structures |

# Benchmarks

## Hardware Specs:
OS: Linux 6.17.7-arch1-1
CPU: AMD Ryzen 7 PRO 4750G (16) @ 4.46 GHz
GPU: AMD Radeon RX 6700 XT
RAM: F4-3200C16D-32GVK

## Basic bench:

The code used to measure can be found [here](https://github.com/Brogolem35/godot/tree/raw_bench). The unit of time used is clock cycles, got with use of `rdtscp`.  `rdtscp` is used instead of `rdtsc`, as `rdtsc` is susceptible to "out-of-order execution", while `rdtscp` is not.

### Results:

Warm access bench (no cache_flush): 
[bench_hot.csv](https://github.com/user-attachments/files/23496329/bench_hot.csv)

Cold access bench (with cache_flush):
[bench_cold.csv](https://github.com/user-attachments/files/23496342/bench_cold.csv)

### My Interpretation:

When it comes to insertion, in almost all cases the AHashSet outperforms HashSet. Sometimes with up to 30% difference.

In warm access, in some cases, there can be seen slight regressions regarding the lookup times, and improvements in other cases. In cold access, the lookup regressions are much more sparse to the point of being exceptions. 

I interpret this improvements are the result of AHashSet using AoS instead of SoA layout for metadata, thus better cold access, and the use of masking instead of fastmod or if checks for wrapping.

## Bunnymark:

[Gaijin Entertainment's Godot Bunnymark](https://github.com/GaijinEntertainment/godot-das/tree/master/examples/bunnymark) with [some changes](https://github.com/Brogolem35/gdscript-bunnymark). Used the release_template build from the [mentioned rawahashtable_all branch](https://github.com/Brogolem35/godot/tree/rawahashtable_all).

### Results:

The number next to HashSet/AHashSet indicates the number of starting bunnies.
[bunnymark.txt](https://github.com/user-attachments/files/23497436/bunnymark.txt)

# Known Issues:

- MSVC build is not working. It would be helpful if someone with experience takes a look.

Brogolem35 committed:
Add RawAHashTable abstract class and base AHashMap on it

Brogolem35 committed:
Add AHashSet

Brogolem35 committed:
Use AHashSet instead of HashSet

Brogolem35 committed:
Use full path to include RawAHashTable

Brogolem35 committed:
Change out of class names to avoid name collision

Brogolem35 committed:
Add some inline hints

Brogolem35 committed:
Reformat

Brogolem35 committed:
Remove inline from overridden methods

Brogolem35 committed:
Supress warnings

Brogolem35 committed:
Merge branch 'master' into rawahashtable

Brogolem35 committed:
Fix warning

Brogolem35 committed:
Update documentation

Brogolem35 committed:
Reformat

Brogolem35 committed:
Remove unused classes in the beginning

AThousandShips commented on core/templates/a_hash_map.h:
```diff
@@ -79,248 +54,115 @@ class Variant;
-		return _lookup_idx_with_hash(p_key, r_element_idx, r_meta_idx, _hash(p_key));
+class AHashMap final : RawAHashTable<TKey, Hasher, Comparator> {
+protected:
+	const TKey &_get_key(uint32_t idx) const override {
```
```suggestion
	const TKey &_get_key(uint32_t p_idx) const override {
```

AThousandShips commented on core/templates/raw/raw_a_hash_table.h:
We don't differentiate other types like `COWData` so I don't think it makes sense to make this more complicated

Brogolem35 replied:
Done.

AThousandShips commented on core/templates/a_hash_map.h:
```diff
@@ -79,248 +54,115 @@ class Variant;
-			uint32_t real_capacity = _capacity_mask + 1;
-			_metadata = reinterpret_cast<Metadata *>(Memory::alloc_static_zeroed(sizeof(Metadata) * real_capacity));
-			_elements = reinterpret_cast<MapKeyValue *>(Memory::alloc_static(sizeof(MapKeyValue) * (_get_resize_count(_capacity_mask) + 1)));
+			uint32_t real_capacity = this->_capacity_mask + 1;
```
No I mean `AHashSet : public RawAHashTable`, as far as I can tell the issue here is that with private inheritance you make the `protected` members of the parent `private`, if you use `public` here all the public members of the parent are public, protected ones remain protected, but with private inheritance they become private, `this` can then be used to override rhis, but it's unnecessary and noisy

Brogolem35 committed:
Change parameter names according to the convention

Brogolem35 committed:
Merge branch 'master' into rawahashtable

Brogolem35 committed:
Add new line at the end

Brogolem35 committed:
Remove `this->` (WIP)

Brogolem35 committed:
Remove `this->`

AThousandShips commented on core/templates/a_hash_set.h:
```diff
@@ -0,0 +1,443 @@
+	using Base::_size;
+
+protected:
+	const TKey &_get_key(uint32_t p_idx) const override {
```
```suggestion
	virtual const TKey &_get_key(uint32_t p_idx) const override {
```

Brogolem35 commented:
> Hi, please make the benchmarks easier to digest. Ideally, i'd like to see graphs (but I'm open to make them for you if you don't know how). And please add some interpretation about them, too. Where do you see improvements? Where do you see regressions? Where do you see risks? And how do you assess the change overall, regarding why it is needed?

I added the relevant benchmarks to the PR itself instead of linking to seperate threads. I also added my specs, my interpretation of the results, and the differences between HashSet and AHashSet.

Brogolem35 commented on core/templates/a_hash_set.h:
```diff
@@ -0,0 +1,443 @@
+	}
+
+	// Inserts an element without checking if it already exists.
+	Iterator insert_new(const TKey &p_key) {
```
I am not sure what you mean by that. Can you be more explicit?

AThousandShips replied:
Does it crash without the check? I.e. is the behavior different between dev and non dev

AThousandShips replied:
Might be good to make a clear note that it is safe to use it "incorrectly" in release builds

Brogolem35 committed:
Update insert_new doc

Brogolem35 committed:
Add virtual attribute to overriden methods

AThousandShips commented:
`HashSet` at least doesn't offer insertion order, not sure about general pointer stability, from the note in the header of that type it seems to not offer pointer stability either on adding/removing elements

AThousandShips commented on core/templates/raw/raw_a_hash_table.h:
```diff
@@ -0,0 +1,232 @@
+
+class String;
+class StringName;
+class Variant;
```
These are not needed here, they were forward declared in `AHashMap` to be able to add the externs at the end

Brogolem35 committed:
Remove unused class declerations

Brogolem35 committed:
Remove _FORCE_INLINE_ from `remove`

Brogolem35 commented on core/templates/a_hash_map.h:
```diff
@@ -552,7 +407,7 @@ class AHashMap {
 	}
 
-	void remove(const Iterator &p_iter) {
+	_FORCE_INLINE_ void remove(const Iterator &p_iter) {
```
Removed.

Brogolem35 committed:
Move RawAHashTable up in the file hierarcy

Brogolem35 commented:
> Since you're effectively re-implementing AHashMap, we need to see benchmarks of the old vs new implementation here as well. Hot benchmarks should suffice, if you didn't change any indirections. The last time I tried abstracting over the hash table implementation, I got some performance regressions, so I am careful with refactors around AHashMap.

Did a hot access bench. The one on the left is old, and the right is new. Because of how CSVs work, I kinda messed up the formatting.
[bench_map.csv](https://github.com/user-attachments/files/23541507/bench_map.csv)

There seems to be a lot of fluctiation in numbers, not sure if it is the result of the code changes or system noise. The regressions seem to be mostly correlate with erase but insert existing is generally improved, sometimes with 20% differences. I am unable to see a clear causation for this.

Brogolem35 committed:
Try fixing MSVC

Brogolem35 committed:
Rename RAHT_* contants and move them to the RawHashTable namespace

Brogolem35 committed:
Extract metadata cleanup to its own method

Brogolem35 committed:
Move some public APIs to RawHashTable

Brogolem35 committed:
Move reset and clear to RawAHashTable

AThousandShips commented on core/templates/raw_a_hash_table.h:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	_FORCE_INLINE_ bool _eq(const TKey &a, const TKey &b) const {
+		return Comparator::compare(a, b);
```
```suggestion
	_FORCE_INLINE_ bool _eq(const TKey &p_a, const TKey &p_b) const {
		return Comparator::compare(p_a, p_b);
```

Brogolem35 committed:
Merge branch 'master' into rawahashtable

Brogolem35 committed:
Merge branch 'master' into rawahashtable

Brogolem35 committed:
Remove extra APIs introduced to the AHashSet

Brogolem35 committed:
Use CRTP to avoid virtualization

Brogolem35 commented on core/templates/raw_a_hash_table.h:
```diff
@@ -0,0 +1,285 @@
+		return (p_meta_idx - original_idx + p_capacity + 1) & p_capacity;
+	}
+
+	_FORCE_INLINE_ virtual const TKey &_get_key(uint32_t p_idx) const = 0;
```
> I can create another PR that does that and close this one if you want.

Brogolem35 replied:
Related: #122544

Brogolem35 committed:
Fix minor grammar error

Brogolem35 committed:
Merge branch 'master' into rawahashtable

Brogolem35 committed:
Reformat
